[SWT-0004] Constrain the granularity of test time limit durations (#534)
swiftlang/swift-testing: Documentation/Proposals/0004-constrain-the-granularity-of-test-time-limit-durations.md

/// more details.
///
/// If a test is parameterized, this time limit is applied to each of its
/// test cases individually. If a test has more than one time limit associated
/// with it, the shortest one is used. A test run may also be configured with
/// a maximum time limit per test case.
public static func timeLimit(_ timeLimit: Self.Duration) -> Self
```

And finally, the call site of the API looks like this:

```swift
@Test(.timeLimit(.minutes(60))
func serve100CustomersInOneHour() async {
  for _ in 0 ..< 100 {
    let customer = await Customer.next()
    await customer.order()
    ...
  }
}
```

The `TimeLimitTrait.Duration` struct has various `unavailable` overloads that
are included for diagnostic purposes only. They are all documented and
annotated like this:

```swift
/// Construct a time limit duration given a number of <unit>.
///
/// This function is unavailable and is provided for diagnostic purposes only.
@available(*, unavailable, message: "Time limit must be specified in minutes")
```

## Source Compatibility

This is purely additional API and does not impact existing code.

## Integration with Supporting Tools

N/A

## Future Directions

> We could allow more finegrained time limits but this was not considered.

Reviewer: ```suggestion
- We could allow more finegrained time limits in the future that scale with the
  performance of the test host device.
```

failure caused by a short timeout can be surprisingly costly, potentially 
leading to an entire CI pipeline being rerun. Determining an appropriate 
timeout for a specific test can be a challenging task.

Additionally, when the system intentionally runs multiple tests simultaneously 
to optimize resource utilization, the scheduler becomes the arbiter of test 
execution. Consequently, the test may take significantly longer than 
anticipated, potentially due to external factors beyond the control of the code 
under test.

A unit test should be capable of failing due to hanging, but it should not fail 
due to being slow, unless the developer has explicitly indicated that it 
should, effectively transforming it into a performance test.

The time limit feature is *not* intended to be used to apply small timeouts to 
tests to ensure test runtime doesn't regress by small amounts. This feature is 
intended to be used to guard against hangs and pathologically long running 
tests.

## Proposed Solution

We propose changing the `.timeLimit` API to accept values of a new `Duration` 
type defined in `TimeLimitTrait` which only allows for `.minute` values to be 
passed.

> These types already exist as SPI and this proposal is seeking to making these 

Reviewer: ```suggestion
This type already exists as SPI and this proposal is seeking to making it
```

API.

## Detailed Design

The `TimeLimitTrait.Duration` struct only has one factory method:
```swift

> public static func minutes(_ minutes: some BinaryInteger) -> Self`

Reviewer: ```suggestion
public static func minutes(_ minutes: some BinaryInteger) -> Self
```

The `TimeLimitTrait.Duration` struct is declared as follows:

```swift
/// A type that defines a time limit to apply to a test.
///
/// To add this trait to a test, use one of the following functions:
///
/// - ``Trait/timeLimit(_:)``
@available(macOS 13.0, iOS 16.0, tvOS 16.0, watchOS 9.0, *)
public struct TimeLimitTrait: TestTrait, SuiteTrait {
  /// A type representing the duration of a time limit applied to a test.
  ///
  /// This type is intended for use specifically for specifying test timeouts
  /// with ``TimeLimitTrait``. It is used instead of Swift's built-in `Duration`
  /// type because test timeouts do not support high-precision, arbitrarily
  /// short durations. The smallest allowed unit of time is minutes.
  public struct Duration: Sendable {

    /// Construct a time limit duration given a number of minutes.
    ///
    /// - Parameters:
    ///   - minutes: The number of minutes the resulting duration should
    ///     represent.
    ///
    /// - Returns: A duration representing the specified number of minutes.
    public static func minutes(_ minutes: some BinaryInteger) -> Self
  }

  /// The maximum amount of time a test may run for before timing out.

>   public var timeLimit: Swift.Duration

Reviewer: ```suggestion
  public var timeLimit: Swift.Duration { get set }
```

/// policy which may be configured on a global basis by the tool responsible
/// for launching the test process. Refer to that tool's documentation for
/// more details.
///
/// If a test is parameterized, this time limit is applied to each of its
/// test cases individually. If a test has more than one time limit associated
/// with it, the shortest one is used. A test run may also be configured with
/// a maximum time limit per test case.
public static func timeLimit(_ timeLimit: Self.Duration) -> Self
```

And finally, the call site of the API looks like this:

```swift
@Test(.timeLimit(.minutes(60))
func serve100CustomersInOneHour() async {
  for _ in 0 ..< 100 {
    let customer = await Customer.next()
    await customer.order()
    ...
  }
}
```

The `TimeLimitTrait.Duration` struct has various `unavailable` overloads that
are included for diagnostic purposes only. They are all documented and
annotated like this:

```swift
/// Construct a time limit duration given a number of <unit>.
///
/// This function is unavailable and is provided for diagnostic purposes only.
@available(*, unavailable, message: "Time limit must be specified in minutes")
```

## Source Compatibility

> This is purely additional API and does not impact existing code.

Reviewer: This is not the case; anyone who has adopted `.timeLimit()` and is passing a `Swift.Duration` instance is impacted.

## Integration with Supporting Tools

N/A

## Future Directions

We could reconsider the granularity constraints and allow for more finegrained
time limits if we come to the conclusion that the advantages outweigh the
disadvantages.

> Part of that could be the automatic detection of environments (like CI vs local

Reviewer: ```suggestion
Part of that could be the automatic detection of environments (like CI vs local)
```

and providing a way to use different timeouts in different environments.

## Alternatives Considered

We have considered using `Swift.Duration` as the currency type for this API but 

> decided against it to avoid common pitfals and misuses of this feature such as 

Reviewer: ```suggestion
decided against it to avoid common pitfalls and misuses of this feature such as 
```

/// test cases individually. If a test has more than one time limit associated
/// with it, the shortest one is used. A test run may also be configured with
/// a maximum time limit per test case.
public static func timeLimit(_ timeLimit: Self.Duration) -> Self
```

And finally, the call site of the API looks like this:

```swift
@Test(.timeLimit(.minutes(60))
func serve100CustomersInOneHour() async {
  for _ in 0 ..< 100 {
    let customer = await Customer.next()
    await customer.order()
    ...
  }
}
```

The `TimeLimitTrait.Duration` struct has various `unavailable` overloads that
are included for diagnostic purposes only. They are all documented and
annotated like this:

```swift
/// Construct a time limit duration given a number of <unit>.
///
/// This function is unavailable and is provided for diagnostic purposes only.
@available(*, unavailable, message: "Time limit must be specified in minutes")
```

## Source Compatibility

This impacts clients that have adopted the `.timeLimit` trait and use overloads
of the trait that accept an arbitrary `Swift.Duration` except if they used the
`minutes` overload.

## Integration with Supporting Tools

N/A

## Future Directions

> We could allow more finegrained time limits in the future that scale with the

Reviewer 2: I think it might be a bit limiting that this constraint is designed into the API.

Unit-test timeouts should often be set to lower than 1 minute.
In my opinion, for most types of testing, 1 minute timeouts are too slow, and a test can reliably be considered to have hang after a few seconds.

Popular unit-testing frameworks on different languages don't seem to impose similar restrictions:

* `mocha` - node / browser
   * 2 seconds - https://mochajs.org/running/cli/
   * 14 million downloads a week https://www.npmjs.com/package/mocha
* `jest` - node / browser
   * 5 seconds - https://jestjs.io/docs/cli#--testtimeoutnumber
   * 47 million downloads a week https://www.npmjs.com/package/jest
* `junit`, `rspec`, `pytest`, `cargo test`
   * no default or no built-in timeouts
   * but also no lower bound - https://docs.junit.org/5.5.0/api/org/junit/jupiter/api/Timeout.html

In general, unit-testing framework timeouts or timeout plug-ins mention timeouts in seconds, not minutes.

- - -

I understand the timeout might have been set here to address integration test runtimes which are longer.
And how in that scenario a timeout set to a low duration could cause problems.

I see the forum discussion had comments:

> As mentioned in the thread, the usage of coarse-grained timeout values is not frequently needed for fast-running unit tests, but can be a very useful backstop for when things go badly wrong especially during integration testing, and being able to declare the expected maximum duration on a per-test/suite basis via traits provides useful flexibility.

But I believe it might be missing something in -

> the usage of coarse-grained timeout values is not frequently needed for fast-running unit tests

The primary reason to add timeouts is to catch stalled tests. If a test hangs, ideally the test-suite will fail.

Ideally a timeout will abort the test-suite _as soon as_ you could reliably detect there's a hang.

In some integration test cases, this might be 1 minute or several minutes. But there are many cases where even a few seconds is enough to reliably know that there's a hang.

For **integration tests**, and in particular tests around asynchronous / concurrent / multi-threaded code, it's nice to be able to detect hangs quickly. On that case, a timeout is useful, and a 1 minute timeout is too high.

- - -
